specs/derivation.md

```diff
@@ -0,0 +1,820 @@
+    - A channel is considered to be *timed out* if `currentL1Block > channeld_id.timestamp + CHANNEL_TIMEOUT`.
+        - where `currentL1Block` is the L1 block maintained by this stage, which is the latest L1 block whose frames
+          have been added to the channel bank.
+
+> **TODO** There is currently `MAX_CHANNEL_BANK_SIZE`, a notion about the maximum amount of channels we can keep track
+> of.
+> - Is this a semantic detail (i.e. if the batcher opens too many frames, valid channels can be dropped?)
+> - If so, I feel **very strongly** about changing this. This ties us very much to the current implementation.
+> - And it doesn't feel necessary given the channel timeout - if DOS is an issue we can reduce the channel timeout.
+
+### Batch Decoding
+
+In the *Batch Decoding* stage, we simply decode [batches][g-sequencer-batch] from the frames we received from the last
+stage.
+
+### Payload Attributes Derivation
+
+In the *Payload Attributes Derivation* stage, we convert the decoded batches into instances of the
+[`PayloadAttributes`][g-payload-attr] structure. Such a structure encodes the transactions that need to figure into a
+block, as well as other block inputs (timestamp, fee recipient, etc). Payload attributes derivation is detailed in the
+section [Deriving Payload Attributes section][deriving-payload-attr] below.
+
+If batches are missing for a given [time slot][g-time-slot], this stage generates empty batches to fill gaps.
+
+### Engine Queue
+
+In the *Engine Queue* stage, the previously derived `PayloadAttributes` structures are buffered and sent to the
+[execution engine][g-exec-engine] to be executed and converted into a proper L2 block.
+
+The engine queue maintains references to two L2 blocks:
+
+- The [safe L2 head][g-safe-l2-head]: everything up to and including this block can be fully derived from the
+  canonical L1 chain.
+- The [unsafe L2 head][g-unsafe-l2-head]: blocks between the safe and unsafe heads are [unsafe
+  blocks][g-unsafe-l2-block] that have not been derived from L1. These blocks either come from sequencing (in sequencer
+  mode) or from [unsafe sync][g-unsafe-sync] to the sequencer (in validator mode).
+
+If the unsafe head is ahead of the safe head, then [consolidation][g-consolidation] is attempted.
+
+During consolidation, we consider the oldest unsafe L2 block, i.e. the unsafe L2 block directly after the safe head. If
+the payload attributes match this oldest unsafe L2 block, then that block can be considered "safe" and becomes the new
+safe head.
+
+In particular, the following fields of the payload attributes are checked for equality with the block:
+
+- `parent_hash`
+- `timestamp`
+- `randao`
+- `transactions_list` (first length, then equality of each of the encoded transactions)
+
+TODO link payload attributes spec / definition
+
+If consolidation fails, the unsafe L2 head is reset to the safe L2 head.
```

Comment:
and then execution proceeds to create a new safe L2 head on top of the old safe L2 head